airbyte-integrations/connectors/source-amazon-seller-partner/source_amazon_seller_partner/streams.py

```diff
@@ -0,0 +1,175 @@
+# SOFTWARE.
+#
+
+from abc import ABC, abstractmethod
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional
+
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+from source_amazon_seller_partner.auth import AWSSignature
+
+REPORTS_API_VERSION = "2020-09-04"
+ORDERS_API_VERSION = "v0"
+
+
+class AmazonSPStream(HttpStream, ABC):
+    page_size = 100
+    data_field = "payload"
+
+    def __init__(self, url_base: str, aws_signature: AWSSignature, replication_start_date: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self._url_base = url_base
+        self._aws_signature = aws_signature
+        self._replication_start_date = replication_start_date
+
+    @property
+    def url_base(self) -> str:
+        return self._url_base
+
+    @property
+    @abstractmethod
+    def replication_start_date_field(self) -> str:
+        pass
+
+    @property
+    @abstractmethod
+    def next_page_token_field(self) -> str:
+        pass
+
+    @property
+    @abstractmethod
+    def page_size_field(self) -> str:
+        pass
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+        if next_page_token:
+            return dict(next_page_token)
+
+        params = {self.replication_start_date_field: self._replication_start_date, self.page_size_field: self.page_size}
+        if self._replication_start_date:
+            start_date = max(stream_state.get(self.cursor_field, self._replication_start_date), self._replication_start_date)
+            params.update({self.replication_start_date_field: start_date})
+        return params
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        stream_data = response.json()
+        next_page_token = stream_data.get(self.next_page_token_field)
+        if next_page_token:
+            return {self.next_page_token_field: next_page_token}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        """
+        :return an iterable containing each record in the response
+        """
+        yield from response.json().get(self.data_field, [])
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        latest_benchmark = latest_record[self.cursor_field]
+        if current_stream_state.get(self.cursor_field):
+            return {self.cursor_field: max(latest_benchmark, current_stream_state[self.cursor_field])}
+        return {self.cursor_field: latest_benchmark}
+
+    def _create_prepared_request(
```

Comment:
can we create an issue to add to the CDK and an issue to update this once that issue is resolved? ideally we should never override an `_` method

Reply:
Sure, here it is https://github.com/airbytehq/airbyte/issues/5246